src/main/java/io/gravitee/policy/apikey/ApiKeyPolicy.java

```diff
@@ -15,138 +15,172 @@
+     *
+     * @return 500
+     */
+    @Override
+    public int order() {
+        return 500;
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        String requestApiKey = lookForApiKey(executionContext, request);
-
-        if (requestApiKey == null || requestApiKey.isEmpty()) {
-            // The api key is required
-            policyChain.failWith(
-                PolicyResult.failure(
-                    API_KEY_MISSING_KEY,
-                    HttpStatusCode.UNAUTHORIZED_401,
-                    "No API Key has been specified in headers (" +
-                    API_KEY_HEADER +
-                    ") or query parameters (" +
-                    API_KEY_QUERY_PARAMETER +
-                    ")."
-                )
-            );
-        } else {
-            try {
-                final String apiId = (String) executionContext.getAttribute(ExecutionContext.ATTR_API);
-
-                Optional<ApiKey> apiKeyOpt = executionContext.getComponent(ApiKeyRepository.class).findByKeyAndApi(requestApiKey, apiId);
-                if (apiKeyOpt.isPresent()) {
-                    ApiKey apiKey = apiKeyOpt.get();
-
-                    // Add data about api-key and subscription into the execution context
-                    executionContext.setAttribute(ExecutionContext.ATTR_APPLICATION, apiKey.getApplication());
-                    executionContext.setAttribute(ExecutionContext.ATTR_SUBSCRIPTION_ID, apiKey.getSubscription());
-                    // Be sure to force the plan to the one linked to the apikey
-                    executionContext.setAttribute(ExecutionContext.ATTR_PLAN, apiKey.getPlan());
-                    executionContext.setAttribute(ATTR_API_KEY, apiKey.getKey());
-
-                    if (
-                        !apiKey.isRevoked() && (apiKey.getExpireAt() == null || apiKey.getExpireAt().after(new Date(request.timestamp())))
-                    ) {
-                        policyChain.doNext(request, response);
-                    } else {
-                        // The api key is not valid
-                        policyChain.failWith(
-                            PolicyResult.failure(
-                                API_KEY_INVALID_KEY,
-                                HttpStatusCode.UNAUTHORIZED_401,
-                                "API Key is not valid or is expired / revoked."
-                            )
+    @Override
+    public Completable onRequest(RequestExecutionContext ctx) {
+        return Completable
+            .defer(() -> {
+                try {
+                    Optional<String> requestApiKey = lookForApiKey(ctx);
+
+                    if (requestApiKey.isEmpty()) {
+                        // The api key is required
+                        return interrupt401(
```

Comment:
Shouldn't be something shared accros secutiry policies to unformized erreur message?

Reply:
I would say that's not a big concern as each policy will use it's own key and default message (same as v3 behavior). There is no real gain to share this for now but definitely something to take care on if we decide to revise the way the SecurityPolicy should handle 401.